pkg/components: add Headlamp component

iaguis opened the pull request:
Fixes https://github.com/kinvolk/lokomotive/issues/988

knrt10 commented on pkg/components/headlamp/component.go:
```diff
@@ -0,0 +1,107 @@
+
+const name = "headlamp"
+
+const chartValuesTmpl = `
```
I think we should follow one format and have `chartValuesTmpl` inside `template.go` file. It's a suggestion. I agree to what the team agrees on.

iaguis replied:
Sure.

knrt10 commented on assets/charts/components/headlamp/Chart.yaml:
```diff
@@ -0,0 +1,7 @@
+type: application
+
+version: 0.1.0
+appVersion: 0.0.0
```
I wonder if this should at least be 0.0.1, even though this is our own chart

invidian replied:
I guess it's `0.0.0` because Headlamp currently do not have any public release (it should though!).

iaguis replied:
It will soon :)

knrt10 commented on assets/charts/components/headlamp/values.yaml:
```diff
@@ -0,0 +1,79 @@
+replicaCount: 1
+
+image:
+  repository: joaquimrocha/headlamp
```
should we push it to `kinvolk/headlamp`?

iaguis replied:
OK. Pushed to quay.io/kinvolk/headlamp

invidian commented on assets/charts/components/headlamp/templates/NOTES.txt:
```diff
@@ -0,0 +1,21 @@
+1. Get the application URL by running these commands:
```
Oh wow, this is hard to read :smile:

invidian replied:
BTW, this won't be visible anyway to the user anyway.

iaguis replied:
It was automatically generated by Helm, I can remove it if you want.

invidian replied:
I usually remove all files from `templates` when creating new chart to keep things simple, if the chart is not intended for external use. This allows keeping things very simple.

invidian commented on assets/charts/components/headlamp/values.yaml:
```diff
@@ -0,0 +1,79 @@
+
+serviceAccount:
+  # Specifies whether a service account should be created
+  create: false
```
If we do not use this value, should we keep it here?

iaguis replied:
Good point.

invidian commented on assets/charts/components/headlamp/templates/deployment.yaml:
```diff
@@ -0,0 +1,65 @@
+          - -plugins-dir=/headlamp/plugins/
+          ports:
+            - name: http
+              containerPort: 4654
```
Perhaps we could deduplicate port into helpers?

iaguis replied:
Will do.

invidian commented on assets/charts/components/headlamp/values.yaml:
```diff
@@ -0,0 +1,79 @@
+
+autoscaling:
+  enabled: false
+  minReplicas: 1
```
This also seems to be not used?

invidian commented on pkg/components/headlamp/component.go:
```diff
@@ -0,0 +1,107 @@
+  enabled: true
+  hosts:
+  - host: {{ .IngressHost }}
+    paths:
```
Isn't that the default?

iaguis replied:
No, by default it's empty list.

invidian commented on pkg/components/headlamp/component.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+type component struct {
+	IngressHost              string `hcl:"ingress_host,attr"`
```
I think we should use `pkg/components/types.Ingress` here for both properties, so we are consistent with other components.

iaguis replied:
Good point.

invidian reviewed (changes requested):
Hmm, I think we should also include headlamp in CI configuration and have some basic e2e tests for it.

iaguis reviewed (commented):
> Hmm, I think we should also include headlamp in CI configuration and have some basic e2e tests for it.

Will add basic tests.

surajssd commented:
@iaguis can you address the reviews. Or do you mind someone taking it over?

invidian commented:
@iaguis make sure to pull the reviewers again when you are done addressing review comments. Resolving existing conversations when they are addressed is also helpful while doing another rounds of reviews :)

iaguis commented:
> @iaguis make sure to pull the reviewers again when you are done addressing review comments. Resolving existing conversations when they are addressed is also helpful while doing another rounds of reviews :)

Yes, I was waiting for CI to pass and I didn't realize it was passing already :D

knrt10 commented on assets/charts/components/headlamp/templates/_helpers.tpl:
```diff
@@ -0,0 +1,62 @@
+{{/*
```
As this is our own chart, I don't think we should use helpers. We can easily set values in templates. We are also not using this in webhook-server.

invidian commented on assets/charts/components/headlamp/templates/deployment.yaml:
```diff
@@ -6,9 +6,6 @@ metadata:
     {{- include "headlamp.labels" . | nindent 4 }}
```
It seems commit 2cbc15ed7e1bc7f5cf294c5974ed2171760667c6 should be split and fixed up into previous comments.

invidian commented on ci/aks/aks-cluster.lokocfg.envsubst:
```diff
@@ -118,3 +118,9 @@ component "httpbin" {
+
+component "headlamp" {
+  ingress {
+    host = "headlamp.${var.cluster_name}.${var.aws_dns_zone}"
```
If we configure ingress on CI, we could also test it that it's reachable.

invidian commented on pkg/components/headlamp/component.go:
```diff
@@ -50,7 +50,15 @@ func (c *component) LoadConfig(configBody *hcl.Body, evalContext *hcl.EvalContex
 	}
 
-	return gohcl.DecodeBody(*configBody, evalContext, c)
+	if err := gohcl.DecodeBody(*configBody, evalContext, c); err != nil {
```
I guess that could be merged with original commit?

invidian commented on pkg/components/headlamp/component.go:
```diff
@@ -0,0 +1,107 @@
+
+	values, err := template.Render(chartValuesTmpl, c)
+	if err != nil {
+		return nil, fmt.Errorf("rendering chart values template failed: %w", err)
```
```suggestion
		return nil, fmt.Errorf("rendering chart values template: %w", err)
```

invidian commented on pkg/components/headlamp/component.go:
```diff
@@ -0,0 +1,107 @@
+
+	renderedFiles, err := util.RenderChart(helmChart, name, c.Metadata().Namespace.Name, values)
+	if err != nil {
+		return nil, fmt.Errorf("rendering chart failed: %w", err)
```
```suggestion
		return nil, fmt.Errorf("rendering chart: %w", err)
```

invidian commented on assets/charts/components/headlamp/templates/deployment.yaml:
```diff
@@ -0,0 +1,61 @@
+  name: {{ include "headlamp.fullname" . }}
+  labels:
+    {{- include "headlamp.labels" . | nindent 4 }}
+  namespace: kube-system
```
Namespace should not be set unless you want this resource to be explicitly created in the namespace other than release namespace.

invidian commented on docs/configuration-reference/components/headlamp.md:
```diff
@@ -0,0 +1,70 @@
+
+Example:
+
+| Argument                                  | Description                                                                                                                                                                 |         Default          |     Type     | Required |
+|-------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------:|:------------:|:--------:|
+| `ingress`                                 | Configuration block for exposing Headlamp through an ingress resource.                                                                                                      |            -             |    block     |   false  |
+| `ingress.host`                            | Used as the `hosts` domain in the ingress resource for headlamp that is automatically created.                                                                              |            -             |    string    |   true   |
+| `ingress.class`                           | Ingress class to use for the Headlamp ingress.                                                                                                                              |        `contour`         |    string    |   false  |
+| `ingress.certmanager_cluster_issuer`      | `ClusterIssuer` to be used by cert-manager while issuing TLS certificates. Supported values: `letsencrypt-production`, `letsencrypt-staging`.                               | `letsencrypt-production` |    string    |   false  |
```
Indentation could be improved with http://markdowntable.com/:
```suggestion
| Argument                             | Description                                                                                                                                   | Default                  | Type   | Required |
|--------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|-:-:----------------------|-:-:----|-:-:------|
| `ingress`                            | Configuration block for exposing Headlamp through an ingress resource.                                                                        | -                        | block  | false    |
| `ingress.host`                       | Used as the `hosts` domain in the ingress resource for headlamp that is automatically created.                                                | -                        | string | true     |
| `ingress.class`                      | Ingress class to use for the Headlamp ingress.                                                                                                | `contour`                | string | false    |
| `ingress.certmanager_cluster_issuer` | `ClusterIssuer` to be used by cert-manager while issuing TLS certificates. Supported values: `letsencrypt-production`, `letsencrypt-staging`. | `letsencrypt-production` | string | false    |

```

invidian commented on pkg/components/headlamp/component.go:
```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 The Lokomotive Authors
```
I'd add unit test for rendering to make sure some manifests are produced.

invidian reviewed (changes requested):
Looks mostly good, though I'd change commits a bit, as right now it is difficult to review the PR commit by commit.

knrt10 reviewed (commented):
NIT picks. We are using `unmarshaling` in our test cases elesewhere.

invidian commented on docs/configuration-reference/components/headlamp.md:
```diff
@@ -0,0 +1,69 @@
+
+## Prerequisites
+
+* A Lokomotive cluster accessible via `kubectl`.
```
Actually, if it's vendor independent, maybe we should say `Kubernetes cluster` here?

invidian commented on docs/configuration-reference/components/headlamp.md:
```diff
@@ -0,0 +1,69 @@
+
+* An ingress controller such as [Contour](contour.md) for HTTP ingress.
+
+* A certificate manager such as [cert-manager](cert-manager.md) for valid certificates.
```
I think we require `cert-manager` specifically here, as we use `cert-manager.io/cluster-issuer` annotation.

invidian commented on docs/configuration-reference/components/headlamp.md:
```diff
@@ -0,0 +1,69 @@
+
+component "headlamp" {
+  ingress {
+    host = "headlamp.example.lokomotive-k8s.org"
+    class = "contour"
+    certmanager_cluster_issuer = "letsencrypt-production"
```
```suggestion
    host                       = "headlamp.example.lokomotive-k8s.org"
    class                      = "contour"
    certmanager_cluster_issuer = "letsencrypt-production"
```

invidian commented on docs/configuration-reference/components/headlamp.md:
```diff
@@ -0,0 +1,69 @@
+|-------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------:|:------------:|:--------:|
+| `ingress.host`                            | Used as the `hosts` domain in the ingress resource for headlamp that is automatically created.                                                                              |            -             |    string    |   true   |
+| `ingress.class`                           | Ingress class to use for the Headlamp ingress.                                                                                                                              |        `contour`         |    string    |   false   |
+| `ingress.certmanager_cluster_issuer`      | `ClusterIssuer` to be used by cert-manager while issuing TLS certificates. Supported values: `letsencrypt-production`, `letsencrypt-staging`.                               | `letsencrypt-production` |    string    |  false   |
```
```suggestion
| Argument                             | Description                                                                                                                                   | Default                  | Type   | Required |
|--------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|-:-:----------------------|-:-:----|-:-:------|
| `ingress.host`                       | Used as the `hosts` domain in the ingress resource for headlamp that is automatically created.                                                | -                        | string | true     |
| `ingress.class`                      | Ingress class to use for the Headlamp ingress.                                                                                                | `contour`                | string | false    |
| `ingress.certmanager_cluster_issuer` | `ClusterIssuer` to be used by cert-manager while issuing TLS certificates. Supported values: `letsencrypt-production`, `letsencrypt-staging`. | `letsencrypt-production` | string | false    |
```

We can remove some whitespace here :)

invidian commented on pkg/components/headlamp/component_test.go:
```diff
@@ -0,0 +1,221 @@
+
+		got, err := ingressFromYAML(m["headlamp/templates/ingress.yaml"])
+		if err != nil {
+			t.Fatalf("Unmarshaling ingress: %v", err)
```
I think it's fine, as we agreed(?) to capitalize error messages in test cases.

knrt10 replied:
cool then

invidian commented on test/ingress/ingress.go:
```diff
@@ -0,0 +1,134 @@
+
+	i, err := client.NetworkingV1beta1().Ingresses(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("getting httpbin ingress: %w", err)
```
Good that you fixed hardcoded name! :joy: 
```suggestion
		return fmt.Errorf("getting %s ingress: %w", name, err)
```

iaguis replied:
oopz :D

invidian commented on test/ingress/ingress.go:
```diff
@@ -0,0 +1,134 @@
+
+// AccessIngress access the first host of Ingress name in namespace in subPath
+// expecting an http StatusOK. Otherwise it returns an error.
+func AccessIngress(t *testing.T, namespace, name, subPath string) error {
```
If we take `t` as an argument here, I _think_ we can just not return error here and use `t.Errorf`/`t.Fatalf`. What do you think?

iaguis replied:
Sure

invidian reviewed (changes requested):
Awesome update @iaguis, I think there is still few bits which could be improved, but it looks very nice already!

invidian commented on go.mod:
```diff
@@ -21,7 +21,7 @@ require (
 	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
 	github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e // indirect
 	github.com/golang/protobuf v1.3.5 // indirect
-	github.com/google/go-cmp v0.5.1 // indirect
```
nit: this should be committed with the code which adds new import, so 3af34fae096718bd628f03ed41faafcad93dc5b9 in this case.

invidian commented on docs/configuration-reference/components/headlamp.md:
```diff
@@ -0,0 +1,70 @@
+
+| Argument                             | Description                                                                                                                                   | Default                  | Type   | Required |
+|--------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|--:-:---------------------|--:-:---|--:-:-----|
+| `namespace`                          | Namespace where Headlampwill be installed.                                                                                                    | "kube-system"            | string | false    |
```
```suggestion
| `namespace`                          | Namespace where Headlamp will be installed.                                                                                                    | "kube-system"            | string | false    |
```

invidian commented on docs/configuration-reference/components/headlamp.md:
```diff
@@ -0,0 +1,70 @@
+| Argument                             | Description                                                                                                                                   | Default                  | Type   | Required |
+|--------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|--:-:---------------------|--:-:---|--:-:-----|
+| `namespace`                          | Namespace where Headlampwill be installed.                                                                                                    | "kube-system"            | string | false    |
+| `ingress`                            | Configuration block for exposing Headlamp through an ingress resource.                                                                        | -                        | block  | false    |
```
I think `Ingress` should be capitalized if we talk about the resource. "ingress" as "incoming traffic" would be fine not-capitalized I guess.
```suggestion
| `ingress`                            | Configuration block for exposing Headlamp through an Ingress resource.                                                                        | -                        | block  | false    |
```

invidian commented on docs/configuration-reference/components/headlamp.md:
```diff
@@ -0,0 +1,70 @@
+
+| Argument                             | Description                                                                                                                                   | Default                  | Type   | Required |
+|--------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|--:-:---------------------|--:-:---|--:-:-----|
+| `namespace`                          | Namespace where Headlampwill be installed.                                                                                                    | "kube-system"            | string | false    |
```
Hm, if headlamp runs in `kube-system`, I guess it's using privileged PSP by default. This combined with exposing via Ingress seems like a potential security weakness.

 Maybe we should move it to `lokomotive-system` instead and add a dedicated PSP for it or run it under restricted one?

iaguis replied:
Moving it to `lokomotive-system` makes sense, I'll give it a go.

invidian commented on pkg/components/headlamp/template.go:
```diff
@@ -0,0 +1,40 @@
+nodeSelector:
+  beta.kubernetes.io/os: linux
+image:
+  # TODO use v0.1.0 when it's released
```
```suggestion
  # TODO: Use v0.1.0 when it's released.
```

invidian commented on pkg/components/headlamp/template.go:
```diff
@@ -0,0 +1,40 @@
+  # TODO use v0.1.0 when it's released
+  tag: l8e
+  pullPolicy: Always
+{{ if .Ingress }}
```
Do we want extra blank line between top level keys?
```suggestion
{{- if .Ingress }}
```

invidian commented on pkg/components/headlamp/template.go:
```diff
@@ -0,0 +1,40 @@
+    kubernetes.io/ingress.class: {{ .Ingress.Class }}
+    cert-manager.io/cluster-issuer: {{ .Ingress.CertManagerClusterIssuer }}
+    contour.heptio.com/websocket-routes: "/"
+{{ end }}
```
nit: It will otherwise have 2 newlines at the end of the content.
```suggestion
{{- end }}
```

invidian commented on test/components/headlamp/headlamp_test.go:
```diff
@@ -0,0 +1,35 @@
+// Copyright 2020 The Lokomotive Authors
```
Commit message:
>It checks that the deployment is created.

Is created and it converges.

invidian commented on test/ingress/aws/aws_test.go:
```diff
@@ -18,129 +18,37 @@
 package aws
```
A lot of happening here in this commit, can we break it down e.g. to bug fix and refactor commits, so it's more clear what has been fixed? Otherwise I only see code moved and it's not obvious from where new values in struct come from.

invidian commented on ci/aks/aks-cluster.lokocfg.envsubst:
```diff
@@ -118,3 +118,9 @@ component "httpbin" {
+
+component "headlamp" {
+  ingress {
+    host = "headlamp.${var.cluster_name}.${var.aws_dns_zone}"
```
Do we want to use production certificates on AKS? If so, why?

iaguis replied:
Oops, good catch.

invidian commented on ci/aws/aws-cluster.lokocfg.envsubst:
```diff
@@ -236,3 +236,10 @@ component "aws-ebs-csi-driver" {}
+
+component "headlamp" {
+  ingress {
+    host = "headlamp.$CLUSTER_ID.$AWS_DNS_ZONE"
+    certmanager_cluster_issuer = "letsencrypt-staging"
```
Attributes should be aligned as with `terraform fmt`.
```suggestion
    host                       = "headlamp.$CLUSTER_ID.$AWS_DNS_ZONE"
    certmanager_cluster_issuer = "letsencrypt-staging"
```

knrt10 commented on test/components/headlamp/headlamp_test.go:
```diff
@@ -0,0 +1,35 @@
+
+const (
+	deploymentName = "headlamp"
+	namespace      = "kube-system"
```
I think this should be `lokomotive-system`

knrt10 commented on test/ingress/headlamp/ingress_test.go:
```diff
@@ -0,0 +1,28 @@
+)
+
+func TestHeadlampIngress(t *testing.T) {
+	testingress.AccessIngress(t, "kube-system", "headlamp", "")
```
same `lokomotive-system`

invidian reviewed (dismissed):
Just one tiny nit, but I'm fine with it.

Nice work @iaguis, LGTM :+1: :heavy_check_mark: :tada:

iaguis commented on pkg/components/headlamp/component_test.go:
```diff
@@ -0,0 +1,221 @@
+		}
+
+		if diff := cmp.Diff(got, want); diff != "" {
+			t.Fatalf("%q: unexpected deployment (-want +got)\n%s", tc.name, diff)
```
Indeed. Fixed.

knrt10 reviewed (approved):
Nice work. Looks great 🔥
